clippy_lints/src/len_zero.rs

```diff
@@ -251,33 +252,133 @@ fn check_trait_items(cx: &LateContext<'_>, visited_trait: &Item<'_>, trait_items
+        let def_id = alias_ty.def_id;
+        let hir_map = cx.tcx.hir();
+        let item = hir_map.get_if_local(def_id);
+        let item = item.unwrap();
+
+        if let Node::Item(item) = item {
+            if let ItemKind::OpaqueTy(opaque) = &item.kind {
+                let bounds = opaque.bounds;
+                if bounds.len() != 1 {
+                    return None;
+                }
+                let bound = &bounds[0];
+                if let GenericBound::LangItemTrait(item, _, _, generic_args) = bound {
+                    if item != &LangItem::Future {
+                        return None;
+                    }
+
+                    let bindings = generic_args.bindings;
+                    if bindings.len() != 1 {
+                        return None;
+                    }
+
+                    let binding = &bindings[0];
+                    let kind = &binding.kind;
+
+                    if let TypeBindingKind::Equality {
+                        term: rustc_hir::Term::Ty(term_ty),
+                    } = kind
+                    {
+                        if let TyKind::Path(QPath::Resolved(_, path)) = &term_ty.kind {
+                            let segments = &path.segments;
+                            if segments.len() != 1 {
+                                return None;
+                            }
+                            let segment = &segments[0];
+                            return Some(segment);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    None
+}
+
+fn is_first_generic_integral<'tcx>(segment: &'tcx PathSegment<'tcx>) -> bool {
+    if let Some(generic_args) = segment.args {
+        if generic_args.args.is_empty() {
+            return false;
+        }
+        let arg = &generic_args.args[0];
+        if let GenericArg::Type(rustc_hir::Ty {
+            kind: TyKind::Path(QPath::Resolved(_, path)),
+            ..
+        }) = arg
+        {
+            let segments = &path.segments;
+            let segment = &segments[0];
+            let res = &segment.res;
+            if matches!(res, Res::PrimTy(PrimTy::Uint(_))) || matches!(res, Res::PrimTy(PrimTy::Int(_))) {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+fn parse_len_output<'tcx>(cx: &LateContext<'tcx>, sig: FnSig<'tcx>) -> Option<LenOutput> {
+    if let Some(segment) = extract_future_output(cx, sig.output()) {
+        let res = segment.res;
+
+        if matches!(res, Res::PrimTy(PrimTy::Uint(_))) || matches!(res, Res::PrimTy(PrimTy::Int(_))) {
+            return Some(LenOutput::Integral);
+        }
+
+        if let Res::Def(_, def_id) = res {
+            if cx.tcx.is_diagnostic_item(sym::Option, def_id) {
+                if is_first_generic_integral(segment) {
+                    return Some(LenOutput::Option(def_id));
+                }
+
+                return Some(LenOutput::Option(def_id));
```

Comment:
Sorry, this is a bug. `is_first_generic_integral` checks that `len()` returns `Option<T>` where T is an integral type. This should ignore things like `pub fn len() -> Option<String>` because that's what happens for the sync version.

I've added a test for this.